simulariumio/data_objects/agent_data.py

```diff
@@ -178,22 +179,34 @@ def get_type_ids_and_mapping(self) -> Tuple[np.ndarray, Dict[str, Any]]:
         last_tid = 0
         for time_index in range(total_steps):
             for agent_index in range(len(self.types[time_index])):
-                tn = self.types[time_index][agent_index]
-                if len(tn) == 0:
+                type_name = self.types[time_index][agent_index]
+                if len(type_name) == 0:
                     continue
-                if tn not in type_id_mapping:
+                if type_name not in type_id_mapping:
                     tid = last_tid
                     last_tid += 1
-                    type_id_mapping[tn] = tid
-                    type_name_mapping[str(tid)] = {"name": tn}
+                    type_id_mapping[type_name] = tid
+                    type_name_mapping[str(tid)] = {"name": type_name}
+                    has_subpoints = self.n_subpoints[time_index][agent_index] > 0
                     if (
-                        tn in self.display_data
-                        and not self.display_data[tn].is_default()
+                        type_name in self.display_data
+                        and not self.display_data[type_name].is_default()
```

Comment:
But even with the above realization I'm confused about this if-else branching... Could you please explain for me what kind of scenario the `if` condition deals with and what it does?? 🙏🏻

Reply:
This is checking whether the display data for this type exists in the mapping and is not just a default display data, i.e. it was provided by the user or engine. If so, we want to use it (and make sure its display type is set), if not, we add a default one so the agent will have a display type